README.md

[![CI Status](https://github.com/ethercrow/lightstep-haskell/workflows/ci/badge.svg)](https://github.com/ethercrow/lightstep-haskell/actions)
![Hackage](https://img.shields.io/hackage/v/lightstep-haskell)

# What is it?

A library for instrumenting your code and sending traces to [LightStep](https://lightstep.com/).


# How to use it?

Example usage:

```haskell
{-# LANGUAGE OverloadedStrings #-}

import Control.Concurrent
import Control.Concurrent.Async
import LightStep.HighLevel.IO (LogEntryKey (..), addLog, currentSpanContext, getEnvConfig, setParentSpanContext, setTag, withSingletonLightStep, withSpan)

seriousBusinessMain :: IO ()
seriousBusinessMain = concurrently_ frontend backend >> threadDelay 1000000
  where
    frontend =
      withSpan "RESTful API" $ do
        threadDelay 10000
        setTag "foo" "bar"
        withSpan "Kafka" $ do
          threadDelay 20000
          setTag "foo" "baz"
        threadDelay 30000
        withSpan "GraphQL" $ do
          threadDelay 40000
          setTag "foo" "quux"
          addLog Event "monkey-job"
          addLog (Custom "foo") "bar"
          withSpan "Mongodb" $ do
            threadDelay 50000
          setTag "lorem" "ipsum"
          threadDelay 60000
        withSpan "data->json" $ pure ()
        withSpan "json->yaml" $ pure ()
        withSpan "yaml->xml" $ pure ()
        withSpan "xml->protobuf" $ pure ()
        withSpan "protobuf->thrift" $ pure ()
        withSpan "thrift->base64" $ pure ()
        threadDelay 70000
    backend =
      withSpan "Background Data Science" $ do
        threadDelay 10000
        withSpan "Tensorflow" $ do
          threadDelay 100000
          setTag "learning" "deep"
        withSpan "Torch" $ do
          threadDelay 100000
          setTag "learning" "very_deep"
        withSpan "Hadoop" $ do
          threadDelay 100000
          setTag "learning" "super_deep"
        withSpan "Parallel map reduce" $ do
          result <- withSpan "Reduce" $ do
            Just ctx <- currentSpanContext
            (a, b) <- do
              threadDelay 100000
              concurrently
                ( withSpan "Calculate a" $ do
                    setParentSpanContext ctx
                    threadDelay 100000
                    return "Lorem "
                )
                ( withSpan "Calculate b" $ do
                    setParentSpanContext ctx
                    threadDelay 100000
                    return "ipsum"
                )
            threadDelay 100000
            pure (a <> b)
          setTag "result" result

main :: IO ()
main = do
  -- Construct a config from env variables
  -- - LIGHTSTEP_ACCESS_TOKEN
  -- - LIGHTSTEP_HOST (optional)
  -- - LIGHTSTEP_PORT (optional)
  -- - LIGHTSTEP_SERVICE (optional)
  Just lsConfig <- getEnvConfig
  withSingletonLightStep lsConfig seriousBusinessMain
  putStrLn "All done"
```


CHANGELOG
Unreleased
  * This package is no longer tested with stack, only with cabal.

0.10.4
  * Tolerate Lightstep being down

0.10.3
  * Track heap allocations per span

0.10.2
  * Change reporter_id from 2 to a random number

0.10.1
  * Fix withSpanAndTags losing tags when it was not a root span

0.10.0
  * Rename withSpanAndSomeInitialTags to withSpanAndTags by popular demand

0.9.0
  * Introduce withTags and withSpanAndSomeInitialTags for more efficient tagging

0.8.0
  * Configurable compression

0.7.0
  * New diagnostic counter: number of sent batches
  * Configurable minimum size of a batch to send to the collector, default is 1 span

0.6.1
  * Removed a forgotten debug print statement

0.6.0
  * Fixed context propagation, changed API a bit, please consult example-req for usage.

0.5.2:
  * Make `modifyCurrentSpan` and `currentSpanContext` functions safe to call not within a span

0.5.1:
  * Trace and span ids are now random instead of timestamps

0.5.0:
  * Introduce getEnvConfig
  * [Highlevel.IO] Auto-tag root spans with a thread id
  * Introduce LightStep.Diagnostics module that exposes some internal metrics that may be of interest when debugging or optimizing this library

0.4.5:
  * [Highlevel.IO] Fix a memory leak of several bytes per instrumented thread

0.4.4:
  * Always include x-b3-sampled=true in B3 context propagation headers

0.4.3:
  * Support B3 context propagation headers (only 64bit trace id)

0.4.2:
  * Make example executables not buildable by default. Use `cabal v2-build -f examples` or `stack build --flag lightstep-haskell:examples` to opt in

0.4.1:
  * Add `lightstep.tracer_version` to global tags

0.4.0:
  * Support GHC-8.8

0.3.0:
  * Introduce Network.Wai.Middleware.LightStep
  * Implement SpanContext (de)serialization into `ot-tracer-{trace,span}id` headers
  * Move examples under examples/ directory now that there is more than one

0.2.0:
  * Automatic reconnects

0.1.7:
  * Introduce timeouts to individual grpc requests

0.1.6:
  * Internal refactoring

0.1.5:
  * [Highlevel.IO] Fixed a deadlock in reporting when the backend is unavailable

0.1.4:
  * [Highlevel.IO] Fixed a deadlock in shutdown sequence

0.1.3:
  * [Highlevel.IO] withSpan now works in transformer stacks instead of only IO

0.1.2:
  * [Highlevel.IO] Automatically set error=true tag if exception is thrown from within withSpan

0.1.1:
  * Graceful shutdown

0.1.0:
  * Initial public release


cli/ExampleIO.hs
{-# LANGUAGE OverloadedStrings #-}

import Control.Concurrent
import Control.Concurrent.Async
import Data.Maybe
import qualified Data.Text as T
import LightStep.HighLevel.IO (LightStepConfig (..), setTag, withSingletonLightStep, withSpan)
import System.Environment
import System.Exit

seriousBusinessMain :: IO ()
seriousBusinessMain = concurrently_ frontend backend
  where
    frontend =
      withSpan "RESTful API" $ do
        threadDelay 10000
        setTag "foo" "bar"
        withSpan "Kafka" $ do
          threadDelay 20000
          setTag "foo" "baz"
        threadDelay 30000
        withSpan "GraphQL" $ do
          threadDelay 40000
          setTag "foo" "quux"
          withSpan "Mongodb" $ do
            threadDelay 50000
          setTag "lorem" "ipsum"
          threadDelay 60000
        withSpan "data->json" $ pure ()
        withSpan "json->yaml" $ pure ()
        withSpan "yaml->xml" $ pure ()
        withSpan "xml->protobuf" $ pure ()
        withSpan "protobuf->thrift" $ pure ()
        withSpan "thrift->base64" $ pure ()
        threadDelay 70000
    backend =
      withSpan "Background Data Science" $ do
        threadDelay 10000
        withSpan "Tensorflow" $ do
          threadDelay 100000
          setTag "learning" "deep"
        withSpan "Torch" $ do
          threadDelay 100000
          setTag "learning" "very_deep"
        withSpan "Hadoop" $ do
          threadDelay 100000
          setTag "learning" "super_deep"

main :: IO ()
main = do
  token <-
    lookupEnv "LIGHTSTEP_TOKEN" >>= \case
      Just t -> pure $ T.pack t
      Nothing -> do
        putStrLn "LIGHTSTEP_TOKEN environment variable not defined"
        exitFailure
  host <- fromMaybe "ingest.lightstep.com" <$> lookupEnv "LIGHTSTEP_HOST"
  port <- maybe 443 read <$> lookupEnv "LIGHTSTEP_PORT"
  let lsConfig = LightStepConfig host port token "helloworld" 5
  withSingletonLightStep lsConfig seriousBusinessMain
  putStrLn "All done"
